### PR TITLE
fix: prefer first non-zero rect for interactability checks

### LINKED ISSUE #1961: "in-view center point" is broken when an inline element contains block-level elements
In chapter 12. Elements, section [12.1 Interactability](https://w3c.github.io/webdriver/#interactability):
The first step of the algorithm to calculate `in-view center point ` says: 
```
1. Let rectangle be the first object of the DOMRect collection returned by calling getClientRects() on element. 
```

That rectangle is then used to find the center of the element.

The problem is when that first object has zero width and height, when another of the returned rects in the collection contain the main part of the element. This happens when an inline element such as a `<a>` contains block-level elements.

When coordinates are not integer, the final `floor` operation will yield coordinates that are outside the element. The `floor` operation doesn't make a lot of sense to me (the user _can_ click on non-integer CSS pixels when the dpi isn't 1), but just removing this `floor` operation wouldn't be good enough: clicks would work, but they wouldn't be at the center of the element.

So I think the step 1. above should be changed to:
```
1. Let rectangle be the first object of the DOMRect collection returned by calling getClientRects() on element, that has non-null width and height. If all have a null width and height, let rectangle be the first object of the returned collection.
```
(maybe this could be extracted to another algorithm)

Here is a test that demonstrates the problem in Firefox:
```python
def test_inline_element_with_zero_size_rects_at_fractional_position(
    session, inline
):
    # An inline <a> wrapping a block-level child produces zero-size line-box
    # rects at the start and end of its getClientRects() list: [0x0, 32x32,
    # 0x0]. Placing the element at a fractional y-coordinate (via
    # margin-top: 0.5px) means Math.floor on the zero-size rect's y lands
    # outside the element and elementsFromPoint misses it.
    # This test verifies that clicking such an element succeeds despite the
    # zero-size first rect, i.e. that the non-zero rect is used to determine
    # the element's position.
    session.url = inline("""
        <style>
          body { margin: 0; }
        </style>
        <div style="margin-top: 0.5px">
          <a id="link" href="#"
             onclick="document.getElementById('log').textContent = 'clicked'">
            <div style="width: 32px; height: 32px;"></div>
          </a>
        </div>
        <div id="log"></div>
    """)
    element = session.find.css("#link", all=False)

    # Verify the layout actually produces zero-size first rect; if this fails
    # the test premise has changed and the test should be revisited.
    first_rect = session.execute_script(
        """
        const r = arguments[0].getClientRects()[0];
        return { w: r.width, h: r.height };
        """,
        args=[element],
    )
    assert first_rect["w"] == 0 and first_rect["h"] == 0, (
        "expected zero-size first rect, got {}x{}".format(
            first_rect["w"], first_rect["h"]
        )
    )

    response = element_click(session, element)
    assert_success(response)

    log = session.find.css("#log", all=False)
    assert log.property("textContent") == "clicked"
```

### PR DESCRIPTION
**Pull request by juliandescottes**
Fixes #1961 

This matches the implementation as updated in Firefox. WPT tests already cover potential edge cases related to this (https://github.com/web-platform-tests/wpt/pull/60371)

- - -
This change is [Reviewable](https://reviewable.io/reviews/w3c/webdriver/1965)

***
<a href="https://pr-preview.s3.amazonaws.com/juliandescottes/webdriver/pull/1965.html" title="Last updated on Jul 8, 2026, 2:27 PM UTC (8e07cfd)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver/1965/926cd31...juliandescottes:8e07cfd.html" title="Last updated on Jul 8, 2026, 2:27 PM UTC (8e07cfd)">Diff</a>

### CODE REVIEW

**Comment by juliandescottes** [comment]
I don't understand the build failure here, it seems to complain about <a>node</a> but this is unrelated to my change?

Edit: No idea why this only started failing now, but let's see: https://github.com/w3c/webdriver/pull/1966

**Comment by whimboo** [comment]
@juliandescottes can you please rebase? CI actions should no longer show this error.

**Comment by juliandescottes** [comment]
> @juliandescottes can you please rebase? CI actions should no longer show this error.

Sure,done!

File: index.html [hunk 1]
```diff
@@ -4591,13 +4591,16 @@ <h3>Interactability</h3>
  is the origin position of the rectangle
  that is the intersection between
  the element&apos;s first {{DOMRect}} of {{Element/getClientRects()}}
+ with non-zero width and height (or the first {{DOMRect}} if none have non-zero dimensions)
  and the <a>initial viewport</a>.
```
Comment (whimboo) [inline index.html:4595]:
```suggestion
<p>An <a>element</a>&apos;s <dfn data-lt="center point">in-view center point</dfn>
is the origin position of the rectangle
that is the intersection between
the element&apos;s {{DOMRect}} and the <a>initial viewport</a>. 
The {{DOMRect}} is the first result of {{Element/getClientRects()}}
with non-zero width and height, or the first result if none qualify.
```

File: index.html [hunk 2]
```diff
@@ -4591,13 +4591,16 @@ <h3>Interactability</h3>
...
  It can be calculated this way:
 
 <ol>
  <li><p>Let <var>rectangle</var> be
-  the first object of the {{DOMRect}} collection
-  returned by calling {{Element/getClientRects()}} on <a><var>element</var></a>.
+  the first object in the {{DOMRect}} collection
+  returned by calling {{Element/getClientRects()}} on <a><var>element</var></a>
+  that has non-zero width and height,
+  or the first object in the collection if no such object is found.
```
Comment (whimboo) [inline index.html:4604]: Another suggestion so that it reads better IMHO. Sadly I cannot get GitHub to let me create a suggestion. :(

> Let rectangle be the first object in the DOMRect collection returned by calling getClientRects() on element whose width and height are both non-zero. If no such object exists, let rectangle be the first object in the collection.

**Review by whimboo — COMMENTED** [review]
Just two wording comments from my side.
